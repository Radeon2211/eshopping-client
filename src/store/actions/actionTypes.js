export const LOGIN_USER = 'LOGIN_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const SET_PLACED_ORDERS = 'SET_PLACED_ORDERS';
export const SET_SELL_HISTORY = 'SET_SELL_HISTORY';
export const SET_PROFILE = 'SET_PROFILE';
export const UPDATE_TRANSACTION = 'UPDATE_TRANSACTION';
export const SET_PRODUCTS = 'SET_PRODUCTS';
export const SET_PRODUCT_DETAILS = 'SET_PRODUCT_DETAILS';
export const DELETE_PRODUCT_DETAILS = 'DELETE_PRODUCT_DETAILS';
export const DELETE_PRODUCT_FROM_LIST = 'DELETE_PRODUCT_FROM_LIST';
export const CLEAR_PRODUCTS = 'CLEAR_PRODUCTS';
export const FORM_START = 'FORM_START';
export const FORM_SUCCESS = 'FORM_SUCCESS';
export const FORM_FAIL = 'FORM_FAIL';
export const DATA_START = 'DATA_START';
export const DATA_SUCCESS = 'DATA_SUCCESS';
export const DATA_FAIL = 'DATA_FAIL';
export const SET_MESSAGE = 'SET_MESSAGE';
export const DELETE_MESSAGE = 'DELETE_MESSAGE';
export const SET_MODAL_OPEN_STATE = 'SET_MODAL_OPEN_STATE';
export const SET_MAX_QUANTITY_PER_PAGE = 'SET_MAX_QUANTITY_PER_PAGE';
