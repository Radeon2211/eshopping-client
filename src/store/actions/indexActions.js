export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions';

export { setModal, deleteMessage, changeProductsPerPage } from './uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions';
