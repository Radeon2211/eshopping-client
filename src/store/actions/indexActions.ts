export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeEmail,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions/authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions/tradeActions';

export { setMessage, setModal, changeProductsPerPage } from './uiActions/uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions/productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions/orderActions';
