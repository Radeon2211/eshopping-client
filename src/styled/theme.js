const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#666',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#00A790',
    greenDark: '#009680',
    red: '#F22613',
  },
  shadows: {
    level1: '0 0 2px rgba(0, 0, 0, .1)',
    level2: '0 0 2px rgba(0, 0, 0, .3)',
    level3: '0 0 3px rgba(0, 0, 0, .3)',
    level4: '0 0 5px rgba(0, 0, 0, .3)',
    level5: '0 0 8px rgba(0, 0, 0, .3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Ubuntu, sans-serif',
    text: 'Lato, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
  },
};

export default theme;
