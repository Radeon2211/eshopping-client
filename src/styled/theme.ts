const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#777',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#06AD96',
    greenDark: '#009680',
    red: '#F22613',
    redLight: '#F94D2A',
    redDark: '#985A5A',
  },
  shadows: {
    level1: '0 0 3px rgba(0, 0, 0, 0.03)',
    level2: '0 0 3px rgba(0, 0, 0, 0.3)',
    level3: '0 0 5px rgba(0, 0, 0, 0.3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Noto Sans, sans-serif',
    text: 'Nunito, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
    level5: 100000,
  },
  fontSizes: {
    level1: '1.3rem',
    level2: '1.4rem',
    level3: '1.5rem',
    level4: '1.7rem',
    level5: '2.2rem',
    level6: '2.6rem',
  },
  durations: {
    level1: 0.1,
    level2: 0.2,
    level3: 0.3,
  },
  lineHeights: {
    level1: '1',
    level2: '1.1',
    level3: '1.2',
    level4: '1.3',
    level5: '1.4',
    level6: '1.5',
  },
};

export type ThemeSpacing = keyof typeof theme.spacings;
export type ThemeZIndex = keyof typeof theme.zIndexes;
export type ThemeFontSize = keyof typeof theme.fontSizes;
export type ThemeDuration = keyof typeof theme.durations;
export type ThemeLineHeight = keyof typeof theme.lineHeights;

export default theme;
