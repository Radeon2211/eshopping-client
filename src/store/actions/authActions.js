import axios from '../../axios';
import * as actionTypes from './actionTypes';
import * as uiActions from './uiActions';
import { getErrorMessage } from '../../shared/utility';
import { DEFAULT_PATH } from '../../shared/constants';

export const setProfile = (profile) => ({
  type: actionTypes.SET_PROFILE,
  profile,
});

export const logout = () => ({
  type: actionTypes.LOGOUT_USER,
});

export const setDeliveryAddress = (deliveryAddress) => ({
  type: actionTypes.SET_DELIVERY_ADDRESS,
  deliveryAddress,
});

export const setOtherUser = (user) => ({
  type: actionTypes.SET_OTHER_USER,
  otherUser: user,
});

export const registerUser = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    const country = creds.country.value;
    const phone = `+${creds.phonePrefix.value} ${creds.phoneNumber}`;
    const contacts = [];
    if (!creds.hideEmail) contacts.push('email');
    if (!creds.hidePhone) contacts.push('phone');
    const correctCreds = {
      ...creds,
      country,
      phone,
      contacts,
    };
    delete correctCreds.hideEmail;
    delete correctCreds.hidePhone;
    delete correctCreds.phoneNumber;
    delete correctCreds.phonePrefix;
    try {
      const { data } = await axios.post('/users', { data: correctCreds });
      dispatch(setProfile(data.user));
      dispatch(uiActions.setAndDeleteMessage('Your account has been created successfully!'));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const loginUser = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { data } = await axios.post('/users/login', { data: creds });
      dispatch(setProfile(data.user));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const fetchProfile = () => {
  return async (dispatch) => {
    try {
      const { data } = await axios.get('/users/me');
      dispatch(setProfile(data.user));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
    } catch (error) {
      dispatch(setProfile(null));
    }
  };
};

export const logoutUser = () => {
  return async (dispatch) => {
    try {
      await axios.post('/users/logout');
      dispatch(setProfile(null));
    } catch (error) {
      dispatch(uiActions.setAndDeleteMessage('Unable to logout. Something went wrong'));
    }
  };
};

export const updateUser = (creds, message) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { data } = await axios.patch('/users/me', creds);
      dispatch(setProfile(data.user));
      dispatch(uiActions.setAndDeleteMessage(message));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const deleteAccount = (creds, history) => {
  return async (dispatch, getState) => {
    dispatch(uiActions.formStart());
    try {
      const { username } = getState().auth.profile;
      await axios.delete('/users/me', { data: creds });
      dispatch(setProfile(null));
      dispatch(
        uiActions.setAndDeleteMessage(`Your account has been deleted. Goodbye ${username}!`),
      );
      dispatch(uiActions.formSuccess());
      history.replace(DEFAULT_PATH);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const changeDeliveryAddress = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { onlyTheseOrders } = creds;
      // eslint-disable-next-line no-param-reassign
      delete creds.onlyTheseOrders;

      if (onlyTheseOrders) {
        dispatch(setDeliveryAddress(creds));
      } else {
        const { data } = await axios.patch('/users/me', creds);
        dispatch(setProfile(data.user));
        dispatch(uiActions.setAndDeleteMessage('Delivery address has been saved in your profile'));
      }
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const fetchOtherUser = (username) => {
  return async (dispatch) => {
    dispatch(uiActions.dataStart());
    try {
      const { data } = await axios.get(`/users/${username}`);
      dispatch(setOtherUser(data.profile));
      dispatch(uiActions.dataEnd());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(setOtherUser(null));
      dispatch(uiActions.dataEnd());
    }
  };
};
