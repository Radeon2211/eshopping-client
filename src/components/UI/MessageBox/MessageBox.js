import React, { useCallback } from 'react';
import styled from 'styled-components';
import { motion, AnimatePresence } from 'framer-motion';
import { useSelector, useDispatch } from 'react-redux';
import theme from '../../../styled/theme';
import * as actions from '../../../store/actions/indexActions';
import MyIcon from '../MyIcon';
import { ReactComponent as PlusIcon } from '../../../images/icons/plus.svg';

export const SC = {};
SC.Wrapper = styled(motion.div)`
  display: flex;
  justify-content: center;
  left: 50%;
  transform: translateX(-50%);
  max-width: 100%;
  padding: ${theme.spacings.level1} ${theme.spacings.level2};
  position: fixed;
  top: ${theme.spacings.level2};
  width: 60rem;
  z-index: ${theme.zIndexes.level5};
`;

SC.Message = styled(motion.div)`
  align-items: center;
  background-color: ${theme.colors.blue};
  box-shadow: ${theme.shadows.level3};
  color: #fff;
  display: flex;
  font-size: ${theme.fontSizes.level2};
  justify-content: space-between;
  margin: 0 ${theme.spacings.level2};
  text-align: justify;
  padding: ${theme.spacings.level1} ${theme.spacings.level2};
  width: 100%;
  word-break: break-word;

  & .close-icon {
    cursor: pointer;
    margin-left: ${theme.spacings.level1};
    transform: rotate(45deg);
  }

  @media only screen and (max-width: 37.5em) {
    font-size: ${theme.fontSizes.level3};
    padding: ${theme.spacings.level2};
  }
`;

export const messageBoxVariants = {
  hidden: {
    pointerEvents: 'none',
    y: '-200%',
  },
  visible: {
    transition: { type: 'spring', duration: theme.durations.level2, stiffness: 110 },
    pointerEvents: 'initial',
    y: 0,
  },
};

const MessageBox = () => {
  const message = useSelector((state) => state.ui.message);
  const dispatch = useDispatch();
  const onSetMessage = useCallback((msg) => dispatch(actions.setMessage(msg)), [dispatch]);

  return (
    <AnimatePresence exitBeforeEnter>
      {message && (
        <SC.Wrapper data-testid="MessageBox">
          <SC.Message
            variants={messageBoxVariants}
            initial="hidden"
            animate="visible"
            exit="hidden"
          >
            <span>{message}</span>
            <MyIcon
              size="small"
              color="#fff"
              onClick={() => onSetMessage('')}
              className="close-icon"
              data-testid="MessageBox-close-icon"
            >
              <PlusIcon />
            </MyIcon>
          </SC.Message>
        </SC.Wrapper>
      )}
    </AnimatePresence>
  );
};

export default MessageBox;
