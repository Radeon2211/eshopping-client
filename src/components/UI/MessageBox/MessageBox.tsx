import { useCallback } from 'react';
import { AnimatePresence } from 'framer-motion';
import * as SC from './MessageBox.sc';
import theme from '../../../styled/theme';
import * as actions from '../../../store/actions/indexActions';
import MyIcon from '../MyIcon';
import { ReactComponent as PlusIcon } from '../../../images/icons/plus.svg';
import { useTypedDispatch, useTypedSelector } from '../../../store/reducers/rootReducer';

export const messageBoxVariants = {
  hidden: {
    y: '-200%',
  },
  visible: {
    transition: { type: 'spring', duration: theme.durations.level2, stiffness: 110 },
    y: 0,
  },
};

export default function MessageBox() {
  const message = useTypedSelector((state) => state.ui.message);
  const dispatch = useTypedDispatch();
  const onSetMessage = useCallback((msg: string) => dispatch(actions.setMessage(msg)), [dispatch]);

  return (
    <AnimatePresence mode="wait">
      {message && (
        <SC.Wrapper data-testid="MessageBox">
          <SC.Message
            variants={messageBoxVariants}
            initial="hidden"
            animate="visible"
            exit="hidden"
          >
            <span>{message}</span>
            <MyIcon
              $size="small"
              $color="#fff"
              onClick={() => onSetMessage('')}
              className="close-icon"
              data-testid="MessageBox-close-icon"
            >
              <PlusIcon />
            </MyIcon>
          </SC.Message>
        </SC.Wrapper>
      )}
    </AnimatePresence>
  );
}
