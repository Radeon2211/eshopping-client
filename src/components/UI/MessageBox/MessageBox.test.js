import React from 'react';
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import theme from '../../../styled/theme';
import MessageBox from './MessageBox';
import * as actions from '../../../store/actions/indexActions';

const mockStore = configureMockStore([thunk]);

const setUp = (message) => {
  const store = mockStore({
    ui: { message },
  });
  store.dispatch = jest.fn();

  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <MessageBox />
        </ThemeProvider>
      </Provider>,
    ),
    store,
  };
};

afterEach(cleanup);

describe('<MessageBox />', () => {
  describe('check how renders', () => {
    it('should render everything correctly if message is NOT empty', () => {
      const { asFragment } = setUp('test message');
      expect(asFragment()).toMatchSnapshot();
    });

    it('should NOT render anything if message is empty', () => {
      const { asFragment } = setUp('');
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('check redux actions calls', () => {
    it('should call setMessage() after close icon click', () => {
      const { store } = setUp('test message');
      expect(store.dispatch).not.toHaveBeenCalled();
      fireEvent.click(screen.getByTestId('MessageBox-close-icon'));
      expect(store.dispatch).toHaveBeenCalledWith(actions.setMessage(''));
    });
  });
});
