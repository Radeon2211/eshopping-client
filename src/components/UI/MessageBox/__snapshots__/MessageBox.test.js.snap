// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageBox /> check how renders should NOT render anything if message is empty 1`] = `<DocumentFragment />`;

exports[`<MessageBox /> check how renders should render everything correctly if message is NOT empty 1`] = `
<DocumentFragment>
  .c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 > svg {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  fill: #fff;
  width: 1.7rem;
  height: 1.7rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  max-width: 100%;
  padding: 0.8rem 1.2rem;
  position: fixed;
  top: 1.2rem;
  width: 60rem;
  z-index: 100000;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #1E8BC3;
  box-shadow: 0 0 5px rgba(0,0,0,0.3);
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.4rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 1.2rem;
  text-align: justify;
  padding: 0.8rem 1.2rem;
  width: 100%;
  word-break: break-word;
}

.c1 .close-icon {
  cursor: pointer;
  margin-left: 0.8rem;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

@media only screen and (max-width:37.5em) {
  .c1 {
    font-size: 1.5rem;
    padding: 1.2rem;
  }
}

<div
    class="c0"
    data-testid="MessageBox"
  >
    <div
      class="c1"
      style="pointer-events: none; transform: translateY(-200%) translateZ(0);"
    >
      <span>
        test message
      </span>
      <div
        class="c2 close-icon"
        color="#fff"
        data-testid="MessageBox-close-icon"
      >
        <svg>
          plus.svg
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
